Allow custom flags in index_format

somini opened the pull request:
So far, only the %S expando is treated correctly in all cases.

For the %Z expando, only the marked and flagged messages are correctly
marked.

somini commented:
As promised on #258, here's a pull request that allows customizing the hardcoded flag characters on the index_format. There's problems with the %Z expando (commit 4085010), I can't understand why the `sprintf` with the three `%s` fails.

flatcap commented:
The `snprintf` with `%s%s%s` in it can be reparsed to:

```c
snprintf (buf2,
    sizeof (buf2),
    "%s%s%s",

    (THREAD_NEW ?
        get_nth_wchar (Flagchars, MUTT_FLAG_NEW_THREAD) :
        (THREAD_OLD ?
            get_nth_wchar (Flagchars, MUTT_FLAG_OLD_THREAD) :
            ((hdr->read && (ctx && ctx->msgnotreadyet != hdr->msgno)) ?
                (hdr->replied ?
                    get_nth_wchar (Flagchars, MUTT_FLAG_REPLIED) :
                    get_nth_wchar (Flagchars, MUTT_FLAG_Z_EMPTY)) :
                (hdr->old ?
                    get_nth_wchar (Flagchars, MUTT_FLAG_OLD) :
                    get_nth_wchar (Flagchars, MUTT_FLAG_NEW))
            )
        )
    ),

    (hdr->deleted ?
        get_nth_wchar (Flagchars, MUTT_FLAG_DELETED) :
        (hdr->attach_del ?
            get_nth_wchar (Flagchars, MUTT_FLAG_DELETED_ATTACH) :
            wch
        )
    ),

    (hdr->tagged ?
        get_nth_wchar (Flagchars, MUTT_FLAG_TAGGED) :
        (hdr->flagged ?
            get_nth_wchar (Flagchars, MUTT_FLAG_IMPORTANT) :
            get_nth_wchar (Tochars, mutt_user_is_recipient (hdr))
        )
    )
);
```

if that makes it any clearer.  I'll think about it in the morning.

flatcap commented:
Hi @somini 
I've just updated your code -- rebased and tweaked to use the new upstream mutt changes.
What exactly isn't working?
Please can you list what you expect to see.

somini commented:
With my code I had duplicated flags on %Z, always reproducible on unread emails.
I can test your version in a few hours,  I'll see if the problem persists.

somini commented:
Well, the problem seems to have been solved, I tested it quickly and it seems to be working.

To expand on my findings:
The commit you rebased worked when I opened this PR, but it only concerned the %S expando. I added the second commit that did the same thing on the %Z expando, but I detected some bugs with non-ASCII characters.
After your rebase, I've added my second commit that changed that particular `sprintf` to the devel branch, since it didn't apply correctly, and the problem went away.

As far as I'm concerned, this is now implemented. What is the documentation that needs updating?

-- 

Here's my personal `flag_chars`. I don't distinguish between threads and messages since I don't usually fold my threads, I don't have that much mail.

    set flag_chars="✪★✗✗R☼☼☼☼  "

somini commented:
Now I see I forgot to edit the commit message, **don't merge** it like that.

flatcap commented:
Hi @somini 
First: the commit I pushed trampled over some of your changes. Sorry about that.
I started again.  I've made a few changes:
- The `#define`'s have become an `enum`
- I've expanded all the ternary operators
- Expanded the commit message

I know that the `?:` mess isn't your fault.  I just thought now would be a good time to clean it up
(they were **so** deeply-nested, it was melting my brain :-)

Please can you give it another test to make sure you're happy.
Also, please can you email me: rich@flatcap.org
Cheers.

somini commented:
Ah, this seems much better now, thanks.

I felt bad about writing all those #define, but it was late and I wanted to see the stars in the index. 
As for the nested ternary operators, I even put those comments because I couldn't understand what each character contained.
